nengo/tests/test_learning_rules.py

```diff
@@ -466,3 +467,113 @@ def test_frozen():
 
     with pytest.raises((ValueError, RuntimeError)):
         a.learning_rate = 1e-1
+
+
+@pytest.mark.parametrize('modifies', ["decoders", "weights"])
+def test_generic_rule(Simulator, nl_nodirect, seed, modifies):
+    n = 100
+    input_vector = np.asarray([0.5, -0.5])
+    dt = 0.001
+    reg = 0.1
+
+    m = nengo.Network(seed=seed)
+    with m:
+        m.config[nengo.Ensemble].neuron_type = nl_nodirect()
+        u = nengo.Node(output=input_vector)
+        a = nengo.Ensemble(n, dimensions=2)
+        b = nengo.Ensemble(n, dimensions=2)
+
+        nengo.Connection(u, a)
+
+        if modifies == "decoders":
+            def regularized_pes(target, data):
+                return -dt / n * np.outer(data[:2], data[2:]) - target * reg
+        else:
+            def regularized_pes(target, data, params):
+                encoders = params[b].encoders
+                gains = params[b].gain
```

Comment:
That's entirely up to the person writing the update function.  You could make a callable class that takes the ensemble as a parameter if you want a more general solution.